presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ExchangeNode.java

```diff
@@ -62,14 +64,17 @@
     // for each source, the list of inputs corresponding to each output
     private final List<List<Symbol>> inputs;
 
+    private final Optional<OrderingScheme> orderingScheme;
+
     @JsonCreator
     public ExchangeNode(
             @JsonProperty("id") PlanNodeId id,
             @JsonProperty("type") Type type,
             @JsonProperty("scope") Scope scope,
             @JsonProperty("partitioningScheme") PartitioningScheme partitioningScheme,
             @JsonProperty("sources") List<PlanNode> sources,
-            @JsonProperty("inputs") List<List<Symbol>> inputs)
+            @JsonProperty("inputs") List<List<Symbol>> inputs,
+            @JsonProperty("orderingScheme") Optional<OrderingScheme> orderingScheme)
```

Comment:
What does `orderingScheme` mean for an exchange node?

Reply:
Short: Combined with `GATHERING` exchange type it means that the data should be merged using provided ordering scheme.

Longer description: In a static Presto plan (e.g: task concurrency, static plan nodes) it's hard to capture dynamic execution properties like number of remote exchange sources. Additionally, we must make sure that at least one row is read from every merge source (otherwise merge will block).
Therefore adding ordering scheme to exchange node seems most natural. You could consider it yet another type of exchange (aka Merge exchange) or property of exchange.

Reply:
@sopel39 I'm still trying to understand this, but it sounds like `orderingScheme` doesn't make sense unless `type` is `GATHERING`. Let's add `checkArgument` for that?